src/server/results.rs

```diff
@@ -0,0 +1,246 @@
+use base64;
+use config::Config;
+use crates::{Crate, GitHubRepo};
+use errors::*;
+use ex::Experiment;
+use report;
+use results::{ReadResults, TestResult};
+use rusoto_core::request::default_tls_client;
+use rusoto_s3::S3Client;
+use serde_json;
+use server::db::{Database, QueryUtils};
+use server::experiments::Experiments;
+use server::tokens::Tokens;
+use std::collections::HashMap;
+use toolchain::Toolchain;
+
+#[derive(Deserialize)]
+pub struct TaskResult {
+    #[serde(rename = "crate")]
+    pub krate: Crate,
+    pub toolchain: Toolchain,
+    pub result: TestResult,
+    pub log: String,
+    pub shas: Vec<(GitHubRepo, String)>,
+}
+
+pub struct ResultsDB<'a> {
+    db: &'a Database,
+}
+
+impl<'a> ResultsDB<'a> {
+    pub fn new(db: &'a Database) -> Self {
+        ResultsDB { db }
+    }
+
+    pub fn store(&self, ex: &Experiment, result: &TaskResult) -> Result<()> {
+        // Ensure the log is valid base64 data
+        // This is a bit wasteful, but there is no `validate` method on the base64 crate
+        base64::decode(&result.log).chain_err(|| "invalid base64 log file provided")?;
```

Comment:
Why not store the log bytes directly in the DB? We save ~30% space, and save the decoding cost when reading logs out.